### PR TITLE
Change kafka table provider properties structure.

### PR DESCRIPTION
**Pull request by dpcollins-google**
This is an intentionally breaking change in the kafka beam SQL table. Currently, using the kafka table provider is impossible with a single reference identifier (LOCATION), and the location field goes entirely unused. This change repurposes the currently unused LOCATION field to be `<single bootstrap broker>/<topic name>`.

This is a breaking change because previously, users could have had LOCATION set to any string they wanted, including those without the <broker>/<topic name> structure, since this field was entirely unused.

This change also changes "bootstrap.servers" which is a comma-separated string in the properties to "bootstrap_servers" which uses a proper array, and makes both the "topics" and "bootstrap_servers" parameters optional (previously, they were actually required, although the documentation said otherwise).

Also update beam documentation to reflect new kafka, pubsub and pubsublite semantics added in this and previous PRs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>SDK</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

**Comment by dpcollins-google** [comment]
R: @robinyqiu
R: @TheNeuralBit

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

**Comment by TheNeuralBit** [comment]
FYI SQL PostCommit is broken at the moment (BEAM-12146). https://github.com/apache/beam/pull/14505 is out to fix.

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java [hunk 1]
```diff
@@ -45,23 +48,57 @@
...
-    List<String> topics = new ArrayList<>(topicsArr.size());
-    for (Object topic : topicsArr) {
-      topics.add(topic.toString());
-    }
+
+    ParsedLocation parsedLocation = parseLocation(checkArgumentNotNull(table.getLocation()));
+    List<String> topics = mergeParam(parsedLocation.topic, properties.getJSONArray("topics"));
+    List<String> allBootstrapServers =
+        mergeParam(parsedLocation.brokerLocation, properties.getJSONArray("bootstrap_servers"));
+    String bootstrapServers = String.join(",", allBootstrapServers);
```
Comment (TheNeuralBit) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:101]: What about making `LOCATION` optional, and only merging it with topics/bootstrap_servers if defined?

File: website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md [hunk 2]
```diff
@@ -473,18 +542,8 @@ Write Mode supports writing to a topic.
 *   CSV (default)
     *   Beam parses the messages, attempting to parse fields according to the
         types specified in the schema.
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md:544]: Is there a jira for making CSV a PayloadSerializer?

File: website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md [hunk 3]
```diff
@@ -427,38 +428,106 @@ TYPE pubsub
...
+*   Pub/Sub Lite supports [Generic Payload Handling](#generic-payload-handling).
+
+### Example
+
+```
+CREATE EXTERNAL TABLE locations (event_timestamp TIMESTAMP, attributes ARRAY<ROW<key VARCHAR, `values` ARRAY<VARBINARY>>>, payload ROW<id INTEGER, location VARCHAR>)
+TYPE pubsublite
+LOCATION 'projects/testing-integration/locations/us-central1-a/topics/user-location'
+```
+
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md:482]: Thanks for updating these docs, looks great!

It looks like there's some trailing whitespace that needs to be removed though (See whitespace precommit)

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java (hunk 1 above)
Comment (dpcollins-google) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:101]: Is LOCATION optional in the beam SQL model? I don't think this makes this non-breaking: the user could still have "kafka://mybroker:8080/brokers" which is similar to what was originally suggested (but unused) for LOCATION, which would be broken (would try to use broker hostport of kafka://mybroker:8080 and topic of brokers, neither of which would be expected).

I can make LOCATION optional (i.e. if empty, ignored), I just don't know what this gets us.

File: website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md (hunk 2 above)
Comment (dpcollins-google) [reply website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md:544]: Not currently. CSV is weird in that one record transforms into multiple rows, which works weird with metadata (it would need to be duplicated, which may or may not make sense for headers if one of them, for example, is used for deduplication further down the line). I also have never heard of users shipping data on Pub/Sub-like systems (I count things like Kafka and Kinesis in this) in CSV format, it tends to be more used with file-based storage.

TLDR: we could, but the semantics would be weird and its probably of minimal value

File: website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md (hunk 3 above)
Comment (dpcollins-google) [reply website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md:482]: Will fix, going to wait until I fix the LOCATION thing if there's anything that needs to be changed.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:101]: I think LOCATION is optional in the DDL. The grammar is defined here, but it's a little opaque to me: https://github.com/apache/beam/blob/8e6a7988765c24169c3872c4cdebb03c3def1d50/sdks/java/extensions/sql/src/main/codegen/includes/parserImpls.ftl#L163-L192

Maybe @apilloud can confirm.

But if that's the case I'm not sure why we have  examples where `LOCATION` is specified and ignored. Why didn't we just create examples that don't specify `LOCATION`?

I think the value in allowing LOCATION to be unspecified would be:
1. backwards compatibility
1. could be more convenient for the multiple bootstrap server and topic case

Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:101]: `LOCATION` is optional, that is indicated by the brackets: `[ ]` When we added DDL we assumed every table would have a single canonical location, which is why we made it a first class field. It appears that assumption may be breaking down.

(@robinyqiu is starting to work on support for ZetaSQL DDL, which doesn't have `LOCATION`. We might end up deprecating the field and moving everything to `TBLPROPERTIES`/`OPTIONS` instead.)

Comment (dpcollins-google) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:107]: Done.

File: website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md (hunk 3 above)
Comment (dpcollins-google) [reply website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md:482]: Done.

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

**Comment by dpcollins-google** [comment]
Run SQL PostCommit

**Review by TheNeuralBit — APPROVED** [review]
LGTM. Do you want to take a look @robinyqiu?

**Comment by robinyqiu** [comment]
I took a very quick look earlier this week and it LGTM. Let me take a closer look now.

**Review by robinyqiu — APPROVED** [review]
This is a breaking change to the syntax. I think we should modify the `changes.md` to make sure an explanation is included in the next release. Otherwise LGTM.

**Comment by TheNeuralBit** [comment]
Ran internal tests on this just to be safe and it looks good. I can merge once CHANGES.md is updated.

**Comment by dpcollins-google** [comment]
> This is a breaking change to the syntax. I think we should modify the `changes.md` to make sure an explanation is included in the next release. Otherwise LGTM.

Done

**Comment by TheNeuralBit** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by dpcollins-google** [comment]
Run Website_Stage_GCS PreCommit
